### DISCUSSION TITLE
Documentation?

### DISCUSSION THREAD

**Discussion by imaperson1060** [fent/node-ytdl-core#1042] [Ideas]
In [the example](https://github.com/fent/node-ytdl-core/blob/master/example/info.json), there are many keys that don't have clear functions. Such as in the videoDetails section, on [line 907](https://github.com/fent/node-ytdl-core/blob/40d0c54da10a12397ac63b0dd39ed27de4ff294b/example/info.json#L907), what is `isUnpluggedCorpus`? Or on [line 1297](https://github.com/fent/node-ytdl-core/blob/40d0c54da10a12397ac63b0dd39ed27de4ff294b/example/info.json#L1297), what is `hasYpcMetadata` (I feel like this should be obvious but I'm missing it). I get that it would be ridiculous to make documentation for a 15,000 line file, but is it possible for some small documentation on the parts people use, like videoDetails?

For example, videoDetails would be:
```ts
"videoDetails": {
   "videoId": string,
   "title": string,
   "lengthSeconds": string,
   "keywords": [
      string
   ],
   "channelId": string,
   "isOwnerViewing": boolean,
   "shortDescription": string,
   "isCrawlable": boolean,
   "thumbnail": {
      "thumbnails": [
         {
            "url": string,
            "width": integer,
            "height": integer
         }
      ]
   },
   "averageRating": double,
   "allowRatings": boolean,
   "viewCount": integer,
   "author": string,
   "isPrivate": boolean,
   "isUnpluggedCorpus": boolean,
   "isLiveContent": boolean
}
```

| Key                           | Type             | What it is                                                   |
| ----------------------------- | ---------------- | ------------------------------------------------------------ |
| videoId                       | string           | The id of the video requested                                |
| title                         | string           | The title of the video requested                             |
| lengthSeconds                 | string           | The amount of seconds in the video requested                 |
| keywords[]                    | list             | A list of keywords defined by the creator of the video       |
| channelId                     | string           | The id of the channel the video is on                        |
| isOwnerViewing                | boolean          | Is the owner the one who requested this video? (only possible with [cookies](https://github.com/fent/node-ytdl-core/blob/master/example/cookies.js)) |
| shortDescription              | string           | The description of the video requested                       |
| isCrawlable                   | boolean          | Can search engines see this video? (only true for public videos) |
| thumbnail                     | object           | An object containing the thumbnail of the video              |
| thumbnail.thumbnails[]        | list             | A list of thumbnails with different resolutions              |
| thumbnail.thumbnails[].url    | string           | The url to download the thumbnail from                       |
| thumbnail.thumbnails[].width  | integer          | The width of the thumbnail                                   |
| thumbnail.thumbnails[].height | integer          | The height of the thumbnail                                  |
| averageRating                 | double (I think) | The like:dislike ratio                                       |
| allowRatings                  | boolean          | Did the creator choose to disable ratings for this video?    |
| viewCount                     | integer          | The amount of views the video has                            |
| author                        | string           | The name of the channel the video was uploaded to            |
| isPrivate                     | boolean          | Is the video private? (leads to the interesting issue where you can't see this when it's true without [cookies](https://github.com/fent/node-ytdl-core/blob/master/example/cookies.js)) |
| isUnpluggedCorpus             | boolean          | ...                                                          |
| isLiveContent                 | boolean          | Is the video requested currently live/premiering?            |

**Comment by jessie-murray** [comment]
Hello,

Did you ever find answers to your questions?

I came across this post searching for `hasYpcMetadata` myself on GitHub, and could not find a definite answer. The best suggestion I've seen was that "YPC" could stand for "YouTube Premium Content" or "YouTube Paid Content". I've only seen it set to `false`, and did not find an example of it being `true` on GitHub. If someone could fetch the metadata on a video that is only accessible to subscribers, it would probably help.

No clue about `isUnpluggedCorpus` though.

Best,
Jessie

**Reply by imaperson1060** [reply]
Thanks for your response! I honestly forgot why I made this post, so I couldn't tell you if I ever found what I was looking for specifically. I don't think I was trying to find the purpose of a property so much as find a property that has the specific information that I was looking for (something that having complete documentation would've probably helped with). Whatever it was, it's been a long time since I needed anything other than just the format URLs.

**Reply by jessie-murray** [reply]
For what it's worth, I've kept looking and am still not 100% sure about the meaning here. At least I finally found a link to a YouTube video that's reserved to channel members, from the tech reviewer MKBHD: https://www.youtube.com/watch?v=2_4ZA5QffCg

But extracting metadata for this video, I was surprised to see the same value `"hasYpcMetadata": false`. How could this be if the video is restricted? What makes it even more confusing is that the letters `Ypc` appear in a section that's specifically about the restricted nature of this video, with `YpcOffer` being part of the key:

```json
    "playabilityStatus": {
        "status": "UNPLAYABLE",
        "reason": "Join this channel to get access to members-only content like this video, and other exclusive perks.",
        "errorScreen": {
            "playerLegacyDesktopYpcOfferRenderer": {
                "itemTitle": "Members-only content",
                "itemThumbnail": "...",
                "offerDescription": "Join this channel to get access to members-only content like this video, and other exclusive perks.",
                "offerId": "sponsors_only_video"
            }
        },
```

So clearly there's a `Ypc`-ish reason why I can't play the video, but it also doesn't have `Ypc` metadata? is it just because I'm not a channel member and so that special members-only metadata is just not available to me? That might be it.

Either way, I think I'm done spending time on this, I just hope someone finds this thread useful some day. Or even better, maybe someone who knows the answer will see it and decide to chime in! One can hope. Having documentation that's not just a collection of reverse-engineering efforts would definitely be helpful.

**Reply by imaperson1060** [reply]
Maybe YPC is referring to movies and tv shows? Though it's weird that YouTube wouldn't consider member only videos to be "paid," if that's what YPC stands for.

**Comment by wujlav** [comment]
isUnpluggedCorpus is a flag indicates that video is from YouTube TV (formerly YouTube Unplugged).
